[charts] Improve SVG `pattern` and `gradient` support (#15720)
mui/mui-x: docs/data/charts/styling/styling.md

From here, you can target any subcomponents with its class name.

{{"demo": "SxStyling.js"}}

### Gradients and patterns

> It is possible to use gradients and patterns as fill for the charts.

Reviewer: ```suggestion
It is possible to use gradients and patterns to fill the charts.
```

> This can be done by using the `sx` prop in conjunction with passing your gradient or pattern definition as children of the chart component.

Reviewer: ```suggestion
This can be done by passing your gradient or pattern definition as children of the chart component.
```

> {{"demo": "GradientBar.js"}}

Reviewer: Do we need this example? Seems better to only advocate for inline gradient such that all elements can use it. especially after moving to HTML legend

Author: Hum, it is probably not needed anymore, since I changed the demo to use `vars` and the color prop, previously I was setting the css class, and the way to do that was very different from pie to the other charts.

> On both of these examples, it is important to note that since the gradient is defined in the parent component, the `sx` prop is used to override the fill property of the child components.
> The way it is done with CSS variables, allows us to use a fallback color in places the gradient is not defined, like in the tooltip.

Reviewer: ```suggestion
Note that the gradient or pattern defined that way is only usable for SVG.
So a direct definition like `color: "url(#Pattern)'` would cause undefined colors in HTML elements such as the tooltip.
The demo solves this issue by using a CSS variable ` '--my-custom-pattern': 'url(#Pattern)'` to specify fallback color with `color: 'var(--my-custom-pattern, #123456)'`.
```
